File: src/app/longRunTelemetry.js
import { getScenarioRuntime } from "../world/scenarios/ScenarioFactory.js";
import { ensureAiRuntimeStats } from "./aiRuntimeStats.js";

function round(value, digits = 2) {
  const safe = Number(value);
  if (!Number.isFinite(safe)) return safe;
  return Number(safe.toFixed(digits));
}

function collectErrorWarnings(state) {
  return (state.metrics.warningLog ?? [])
    .filter((entry) => String(entry.level ?? "").toLowerCase() === "error")
    .map((entry) => ({
      sec: round(entry.sec ?? 0, 1),
      source: String(entry.source ?? ""),
      message: String(entry.message ?? "").trim(),
    }))
    .filter((entry) => entry.message);
}

function summarizeWeather(state) {
  const fronts = Array.isArray(state.weather.hazardFronts) ? state.weather.hazardFronts : [];
  return {
    current: String(state.weather.current ?? "clear"),
    timeLeftSec: round(state.weather.timeLeftSec ?? 0, 1),
    pressureScore: round(state.weather.pressureScore ?? 0, 2),
    hazardFrontCount: fronts.length,
    hazardFocusSummary: String(state.weather.hazardFocusSummary ?? ""),
    summary: fronts.length > 0
      ? `Weather: ${state.weather.current}, ${fronts.length} hazard fronts, focus ${state.weather.hazardFocusSummary || "n/a"}`
      : `Weather: ${state.weather.current}`,
  };
}

function collectNonFiniteMetrics(sample) {
  const candidates = [
    ["tick", sample.tick],
    ["simTimeSec", sample.simTimeSec],
    ["resources.food", sample.resources.food],
    ["resources.wood", sample.resources.wood],
    ["gameplay.prosperity", sample.gameplay.prosperity],
    ["gameplay.threat", sample.gameplay.threat],
    ["performance.fps", sample.performance.fps],
    ["performance.frameMs", sample.performance.frameMs],
    ["performance.heapMb", sample.performance.heapMb],
    ["ai.avgLatencyMs", sample.ai.avgLatencyMs],
  ];
  return candidates
    .filter(([, value]) => !Number.isFinite(Number(value)))
    .map(([key]) => key);
}

function buildPopulationByGroup(state) {
  const agents = Array.isArray(state.agents) ? state.agents.filter((entity) => entity?.alive !== false) : [];
  const animals = Array.isArray(state.animals) ? state.animals.filter((entity) => entity?.alive !== false) : [];
  return {
    workers: agents.filter((entity) => entity.type === "WORKER").length,
    traders: agents.filter((entity) => entity.type === "VISITOR" && String(entity.kind ?? "") === "TRADER").length,
    saboteurs: agents.filter((entity) => entity.type === "VISITOR" && String(entity.kind ?? "") !== "TRADER").length,
    herbivores: animals.filter((entity) => entity.kind === "HERBIVORE").length,
    predators: animals.filter((entity) => entity.kind === "PREDATOR").length,
  };
}

function buildEcologyZoneStats(state) {
  const scenario = state.gameplay?.scenario ?? {};
  const ecology = state.metrics?.ecology ?? {};
  const explicit = Array.isArray(ecology.zoneStats) ? ecology.zoneStats : [];
  if (explicit.length > 0) {
    return explicit.map((entry) => ({
      id: String(entry.id ?? ""),
      label: String(entry.label ?? ""),
      herbivoreCount: Number(entry.herbivoreCount ?? 0),
      predatorCount: Number(entry.predatorCount ?? 0),
      herbivoreCapacity: { ...(entry.herbivoreCapacity ?? {}) },
      predatorCapacity: { ...(entry.predatorCapacity ?? {}) },
      recoveryCooldownSec: round(entry.recoveryCooldownSec ?? 0, 2),
      breedingCooldownSec: round(entry.breedingCooldownSec ?? 0, 2),
      predatorRecoveryCooldownSec: round(entry.predatorRecoveryCooldownSec ?? 0, 2),
      herbivoreLowSec: round(entry.herbivoreLowSec ?? 0, 2),
      predatorAbsentSec: round(entry.predatorAbsentSec ?? 0, 2),
      stableSec: round(entry.stableSec ?? 0, 2),
      extinctionSec: round(entry.extinctionSec ?? 0, 2),
      crowdScore: round(entry.crowdScore ?? 0, 2),
    }));
  }
  return (scenario.wildlifeZones ?? []).map((zone) => ({
    id: String(zone.id ?? ""),
    label: String(zone.label ?? ""),
    herbivoreCount: Number(ecology.herbivoresByZone?.[zone.id] ?? 0),
    predatorCount: Number(ecology.predatorsByZone?.[zone.id] ?? 0),
    herbivoreCapacity: {},
    predatorCapacity: {},
    recoveryCooldownSec: 0,
    breedingCooldownSec: 0,
    predatorRecoveryCooldownSec: 0,
    herbivoreLowSec: 0,
    predatorAbsentSec: 0,
    stableSec: 0,
    extinctionSec: 0,
    crowdScore: 0,
  }));
}

export function buildLongRunTelemetry(state, viewState = null) {
  const runtime = getScenarioRuntime(state);
  const objective = state.gameplay?.objectives?.[state.gameplay?.objectiveIndex ?? 0] ?? null;
  const aiRuntime = ensureAiRuntimeStats(state);
  const errorWarnings = collectErrorWarnings(state);
  const populationByGroup = buildPopulationByGroup(state);
  const ecologyZoneStats = buildEcologyZoneStats(state);
  const telemetry = {
    capturedAtIso: new Date().toISOString(),
    phase: String(state.session?.phase ?? "menu"),
    tick: Number(state.metrics?.tick ?? 0),
    simTimeSec: round(state.metrics?.timeSec ?? 0, 2),
    world: {
      templateId: String(state.world?.mapTemplateId ?? ""),
      templateName: String(state.world?.mapTemplateName ?? ""),
      scenarioId: String(runtime.scenario?.id ?? ""),
      scenarioTitle: String(runtime.scenario?.title ?? ""),
      scenarioFamily: String(runtime.scenario?.family ?? ""),
      buildings: { ...runtime.counts },
      frontier: {
        connectedRoutes: Number(runtime.connectedRoutes ?? 0),
        totalRoutes: Number(runtime.routes?.length ?? 0),
        readyDepots: Number(runtime.readyDepots ?? 0),
        totalDepots: Number(runtime.depots?.length ?? 0),
        brokenRoutes: (runtime.routes ?? []).filter((route) => !route.connected).map((route) => route.label).slice(0, 3),
        unreadyDepots: (runtime.depots ?? []).filter((depot) => !depot.ready).map((depot) => depot.label).slice(0, 3),
      },
      weather: summarizeWeather(state),
      events: (state.events?.active ?? []).map((event) => ({
        type: String(event.type ?? ""),
        status: String(event.status ?? ""),
        intensity: round(event.intensity ?? 0, 2),
        targetLabel: String(event.payload?.targetLabel ?? ""),
        pressure: round(event.payload?.pressure ?? event.intensity ?? 0, 2),
        contestedTiles: Number(event.payload?.contestedTiles ?? 0),
      })),
      spatialPressure: {
        weatherPressure: round(state.metrics?.spatialPressure?.weatherPressure ?? 0, 2),
        eventPressure: round(state.metrics?.spatialPressure?.eventPressure ?? 0, 2),
        contestedZones: Number(state.metrics?.spatialPressure?.contestedZones ?? 0),
        contestedTiles: Number(state.metrics?.spatialPressure?.contestedTiles ?? 0),
        activeEventCount: Number(state.metrics?.spatialPressure?.activeEventCount ?? 0),
        peakEventSeverity: round(state.metrics?.spatialPressure?.peakEventSeverity ?? 0, 2),
        summary: String(state.metrics?.spatialPressure?.summary ?? "Spatial pressure: idle"),
      },
    },
    objective: {
      index: Number(state.gameplay?.objectiveIndex ?? 0),
      id: String(objective?.id ?? ""),
      title: String(objective?.title ?? ""),
      progress: round(objective?.progress ?? 100, 1),
      hint: String(state.gameplay?.objectiveHint ?? ""),
    },
    gameplay: {
      prosperity: round(state.gameplay?.prosperity ?? 0, 2),
      threat: round(state.gameplay?.threat ?? 0, 2),
      recovery: {
        charges: Number(state.gameplay?.recovery?.charges ?? 0),
        activeBoostSec: round(state.gameplay?.recovery?.activeBoostSec ?? 0, 1),
        collapseRisk: round(state.gameplay?.recovery?.collapseRisk ?? 0, 1),
        lastReason: String(state.gameplay?.recovery?.lastReason ?? ""),
      },
    },
    resources: {
      food: round(state.resources?.food ?? 0, 2),
      wood: round(state.resources?.wood ?? 0, 2),
    },
    population: {
      byGroup: populationByGroup,
    },
    deaths: {
      total: Number(state.metrics?.deathsTotal ?? 0),
      byReason: { ...(state.metrics?.deathsByReason ?? {}) },
    },
    logistics: {
      carryingWorkers: Number(state.metrics?.logistics?.carryingWorkers ?? 0),
      totalCarryInTransit: round(state.metrics?.logistics?.totalCarryInTransit ?? 0, 2),
      avgDepotDistance: round(state.metrics?.logistics?.avgDepotDistance ?? 0, 2),
      strandedCarryWorkers: Number(state.metrics?.logistics?.strandedCarryWorkers ?? 0),
      overloadedWarehouses: Number(state.metrics?.logistics?.overloadedWarehouses ?? 0),
      busiestWarehouseLoad: Number(state.metrics?.logistics?.busiestWarehouseLoad ?? 0),
      stretchedWorksites: Number(state.metrics?.logistics?.stretchedWorksites ?? 0),
      isolatedWorksites: Number(state.metrics?.logistics?.isolatedWorksites ?? 0),
      warehouseLoadByKey: { ...(state.metrics?.logistics?.warehouseLoadByKey ?? {}) },
      summary: String(state.metrics?.logistics?.summary ?? "Logistics: idle"),
    },
    ecology: {
      activeGrazers: Number(state.metrics?.ecology?.activeGrazers ?? 0),
      pressuredFarms: Number(state.metrics?.ecology?.pressuredFarms ?? 0),
      maxFarmPressure: round(state.metrics?.ecology?.maxFarmPressure ?? 0, 2),
      frontierPredators: Number(state.metrics?.ecology?.frontierPredators ?? 0),
      migrationHerds: Number(state.metrics?.ecology?.migrationHerds ?? 0),
      hotspotFarms: [...(state.metrics?.ecology?.hotspotFarms ?? [])],
      zoneStats: ecologyZoneStats,
      events: {
        ...(state.metrics?.ecology?.events ?? {}),
      },
      clusters: {
        ...(state.metrics?.ecology?.clusters ?? {}),
      },
      flags: {
        ...(state.metrics?.ecology?.flags ?? {}),
      },
      summary: String(state.metrics?.ecology?.summary ?? "Ecology: idle"),
      activePressure: round(
        Math.max(
          Number(state.metrics?.ecology?.maxFarmPressure ?? 0),
          Number(state.metrics?.spatialPressure?.eventPressure ?? 0),
          Number(state.metrics?.spatialPressure?.weatherPressure ?? 0),
        ),
        2,
      ),
    },
    warnings: {
      count: Number((state.metrics?.warningLog ?? []).length),
      errorCount: errorWarnings.length,
      errorWarnings,
    },
    performance: {
      fps: round(state.metrics?.averageFps ?? 0, 2),
      frameMs: round(state.metrics?.frameMs ?? 0, 2),
      headroomFps: round(
        Number(state.metrics?.frameMs ?? 0) > 0
          ? 1000 / Math.max(0.0001, Number(state.metrics.frameMs))
          : 0,
        2,
      ),
      heapMb: round(state.metrics?.memoryMb ?? 0, 2),
      renderMode: String(state.debug?.renderMode ?? "unknown"),
      entityCount: Number(state.metrics?.populationStats?.totalEntities ?? (state.agents.length + state.animals.length)),
      renderFrameCount: Number(state.metrics?.renderFrameCount ?? 0),
      simStepsThisFrame: Number(state.metrics?.simStepsThisFrame ?? 0),
      uiCpuMs: round(state.metrics?.uiCpuMs ?? 0, 2),
      renderCpuMs: round(state.metrics?.renderCpuMs ?? 0, 2),
    },
    ai: {
      enabled: Boolean(state.ai?.enabled),
      coverageTarget: String(state.ai?.coverageTarget ?? "fallback"),
      mode: String(state.ai?.mode ?? "fallback"),
      runtimeProfile: String(state.ai?.runtimeProfile ?? "default"),
      model: String(state.ai?.lastPolicyModel || state.ai?.lastEnvironmentModel || state.metrics?.proxyModel || ""),
      latencyMs: round(state.metrics?.aiLatencyMs ?? 0, 2),
      error: String(state.ai?.lastError || state.ai?.lastPolicyError || state.ai?.lastEnvironmentError || ""),
      proxyHealth: String(state.metrics?.proxyHealth ?? "unknown"),
      fallbackActive: String(state.ai?.mode ?? "fallback") !== "llm",
      requestCount: Number(aiRuntime.requestCount ?? 0),
      timeoutCount: Number(aiRuntime.timeoutCount ?? 0),
      fallbackCount: Number(aiRuntime.fallbackResponseCount ?? 0),
      llmCount: Number(aiRuntime.llmResponseCount ?? 0),
      avgLatencyMs: round(aiRuntime.avgLatencyMs ?? 0, 2),
      consecutiveFallbackResponses: Number(aiRuntime.consecutiveFallbackResponses ?? 0),
      maxUnrecoveredFallbackSec: round(aiRuntime.maxUnrecoveredFallbackSec ?? 0, 2),
      recoveryCount: Number(aiRuntime.recoveryCount ?? 0),
      lastErrorKind: String(aiRuntime.lastErrorKind ?? "none"),
      liveCoverageSatisfied: Boolean(aiRuntime.liveCoverageSatisfied),
    },
    view: {
      targetX: round(viewState?.targetX ?? 0, 2),
      targetZ: round(viewState?.targetZ ?? 0, 2),
      zoom: round(viewState?.zoom ?? 0, 3),
    },
    actionMessage: String(state.controls?.actionMessage ?? ""),
  };
  telemetry.nonFiniteMetrics = collectNonFiniteMetrics(telemetry);
  return telemetry;
}

File: project_utopia/app/long_run_telemetry.py
"""Long-run telemetry builder (port of ``src/app/longRunTelemetry.js``).

Produces a single JSON-shaped sample summarizing a long-horizon simulation
run — world template, scenario, gameplay metrics, AI runtime stats, etc.
The output is consumed by NDJSON post-processors and the future paper
runner ``scripts/benchmark-paper.mjs`` (and its Python sibling).

Dependencies on :mod:`project_utopia.world.scenarios` are imported lazily
because that subagent hasn't landed yet; we fall back to reading
``state.gameplay.scenario`` directly when the runtime helper is absent.
"""

from __future__ import annotations

import math
from datetime import datetime, timezone
from typing import Any

from .ai_runtime_stats import ensure_ai_runtime_stats

__all__ = ["build_long_run_telemetry"]


def _round(value: Any, digits: int = 2) -> float:
    try:
        n = float(value)
    except (TypeError, ValueError):
        return 0.0
    if not math.isfinite(n):
        return n
    return round(n, digits)


def _get_scenario_runtime(state: dict[str, Any]) -> dict[str, Any]:
    """Read scenario runtime info, falling back to ``gameplay.scenario``.

    The JS source uses ``getScenarioRuntime(state)`` which lives in
    :mod:`world.scenarios.ScenarioFactory`. That module is ported by a
    different subagent; until it lands we synthesize a minimal runtime
    block from the state directly.
    """
    try:
        from project_utopia.world.scenarios import get_scenario_runtime  # type: ignore[import-not-found]
    except (ImportError, ModuleNotFoundError):
        gameplay = state.get("gameplay") or {}
        scenario = gameplay.get("scenario") or {}
        return {
            "scenario": scenario,
            "counts": {},
            "routes": [],
            "depots": [],
            "connectedRoutes": 0,
            "readyDepots": 0,
        }
    return get_scenario_runtime(state)


def _collect_error_warnings(state: dict[str, Any]) -> list[dict[str, Any]]:
    metrics = state.get("metrics") or {}
    warning_log = metrics.get("warningLog") or []
    out: list[dict[str, Any]] = []
    for entry in warning_log:
        if not isinstance(entry, dict):
            continue
        if str(entry.get("level", "")).lower() != "error":
            continue
        message = str(entry.get("message", "")).strip()
        if not message:
            continue
        out.append(
            {
                "sec": _round(entry.get("sec", 0), 1),
                "source": str(entry.get("source", "")),
                "message": message,
            }
        )
    return out


def _summarize_weather(state: dict[str, Any]) -> dict[str, Any]:
    weather = state.get("weather") or {}
    fronts = weather.get("hazardFronts") if isinstance(weather.get("hazardFronts"), list) else []
    current = str(weather.get("current", "clear"))
    return {
        "current": current,
        "timeLeftSec": _round(weather.get("timeLeftSec", 0), 1),
        "pressureScore": _round(weather.get("pressureScore", 0), 2),
        "hazardFrontCount": len(fronts),
        "hazardFocusSummary": str(weather.get("hazardFocusSummary", "")),
        "summary": (
            f"Weather: {current}, {len(fronts)} hazard fronts, "
            f"focus {weather.get('hazardFocusSummary') or 'n/a'}"
            if fronts
            else f"Weather: {current}"
        ),
    }


def _collect_non_finite_metrics(sample: dict[str, Any]) -> list[str]:
    candidates: list[tuple[str, Any]] = [
        ("tick", sample.get("tick")),
        ("simTimeSec", sample.get("simTimeSec")),
        ("resources.food", sample.get("resources", {}).get("food")),
        ("resources.wood", sample.get("resources", {}).get("wood")),
        ("gameplay.prosperity", sample.get("gameplay", {}).get("prosperity")),
        ("gameplay.threat", sample.get("gameplay", {}).get("threat")),
        ("performance.fps", sample.get("performance", {}).get("fps")),
        ("performance.frameMs", sample.get("performance", {}).get("frameMs")),
        ("performance.heapMb", sample.get("performance", {}).get("heapMb")),
        ("ai.avgLatencyMs", sample.get("ai", {}).get("avgLatencyMs")),
    ]
    bad: list[str] = []
    for key, value in candidates:
        try:
            n = float(value)
        except (TypeError, ValueError):
            bad.append(key)
            continue
        if not math.isfinite(n):
            bad.append(key)
    return bad


def _build_top_system_ms(state: dict[str, Any], limit: int = 3) -> list[dict[str, Any]]:
    debug = state.get("debug") or {}
    timings = debug.get("systemTimingsMs")
    if not isinstance(timings, dict):
        return []
    entries: list[dict[str, Any]] = []
    for name, stat in timings.items():
        if not isinstance(stat, dict):
            continue
        entries.append(
            {
                "name": str(name),
                "last": _round(stat.get("last", 0), 2),
                "avg": _round(stat.get("avg", 0), 2),
                "peak": _round(stat.get("peak", 0), 2),
            }
        )
    entries.sort(key=lambda e: (-float(e["peak"]), -float(e["avg"])))
    return entries[:limit]


def _build_population_by_group(state: dict[str, Any]) -> dict[str, int]:
    agents = state.get("agents") or []
    animals = state.get("animals") or []
    live_agents = [a for a in agents if isinstance(a, dict) and a.get("alive") is not False]
    live_animals = [a for a in animals if isinstance(a, dict) and a.get("alive") is not False]
    return {
        "workers": sum(1 for a in live_agents if a.get("type") == "WORKER"),
        "traders": sum(
            1
            for a in live_agents
            if a.get("type") == "VISITOR" and str(a.get("kind", "")) == "TRADER"
        ),
        "saboteurs": sum(
            1
            for a in live_agents
            if a.get("type") == "VISITOR" and str(a.get("kind", "")) != "TRADER"
        ),
        "herbivores": sum(1 for a in live_animals if a.get("kind") == "HERBIVORE"),
        "predators": sum(1 for a in live_animals if a.get("kind") == "PREDATOR"),
    }


def build_long_run_telemetry(
    state: dict[str, Any],
    view_state: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Build a single long-run telemetry sample.

    Parameters
    ----------
    state
        Current full game state.
    view_state
        Optional camera state ``{targetX, targetZ, zoom}``. Phase-1
        callers may omit this — the harness has no camera.

    Returns
    -------
    dict[str, Any]
        JSON-serializable dict with the same shape as the JS source.
    """
    runtime = _get_scenario_runtime(state)
    gameplay = state.get("gameplay") or {}
    objectives = gameplay.get("objectives") or []
    obj_index = int(gameplay.get("objectiveIndex", 0) or 0)
    objective = objectives[obj_index] if 0 <= obj_index < len(objectives) else None

    ai_runtime = ensure_ai_runtime_stats(state)
    error_warnings = _collect_error_warnings(state)
    population_by_group = _build_population_by_group(state)

    metrics = state.get("metrics") or {}
    spatial = metrics.get("spatialPressure") or {}
    logistics = metrics.get("logistics") or {}
    ecology = metrics.get("ecology") or {}
    deaths = metrics.get("deathsByReason") or {}

    telemetry: dict[str, Any] = {
        "capturedAtIso": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "phase": str((state.get("session") or {}).get("phase", "menu")),
        "tick": int(metrics.get("tick", 0) or 0),
        "simTimeSec": _round(metrics.get("timeSec", 0), 2),
        "world": {
            "templateId": str((state.get("world") or {}).get("mapTemplateId", "")),
            "templateName": str((state.get("world") or {}).get("mapTemplateName", "")),
            "scenarioId": str((runtime.get("scenario") or {}).get("id", "")),
            "scenarioTitle": str((runtime.get("scenario") or {}).get("title", "")),
            "scenarioFamily": str((runtime.get("scenario") or {}).get("family", "")),
            "buildings": dict(runtime.get("counts", {}) or {}),
            "frontier": {
                "connectedRoutes": int(runtime.get("connectedRoutes", 0) or 0),
                "totalRoutes": len(runtime.get("routes") or []),
                "readyDepots": int(runtime.get("readyDepots", 0) or 0),
                "totalDepots": len(runtime.get("depots") or []),
                "brokenRoutes": [
                    str(r.get("label", ""))
                    for r in (runtime.get("routes") or [])
                    if isinstance(r, dict) and not r.get("connected")
                ][:3],
                "unreadyDepots": [
                    str(d.get("label", ""))
                    for d in (runtime.get("depots") or [])
                    if isinstance(d, dict) and not d.get("ready")
                ][:3],
            },
            "weather": _summarize_weather(state),
            "events": [
                {
                    "type": str(ev.get("type", "")),
                    "status": str(ev.get("status", "")),
                    "intensity": _round(ev.get("intensity", 0), 2),
                    "targetLabel": str((ev.get("payload") or {}).get("targetLabel", "")),
                    "pressure": _round(
                        (ev.get("payload") or {}).get("pressure", ev.get("intensity", 0)), 2
                    ),
                    "contestedTiles": int((ev.get("payload") or {}).get("contestedTiles", 0) or 0),
                }
                for ev in ((state.get("events") or {}).get("active") or [])
                if isinstance(ev, dict)
            ],
            "spatialPressure": {
                "weatherPressure": _round(spatial.get("weatherPressure", 0), 2),
                "eventPressure": _round(spatial.get("eventPressure", 0), 2),
                "contestedZones": int(spatial.get("contestedZones", 0) or 0),
                "contestedTiles": int(spatial.get("contestedTiles", 0) or 0),
                "activeEventCount": int(spatial.get("activeEventCount", 0) or 0),
                "peakEventSeverity": _round(spatial.get("peakEventSeverity", 0), 2),
                "summary": str(spatial.get("summary", "Spatial pressure: idle")),
            },
        },
        "objective": {
            "index": obj_index,
            "id": str((objective or {}).get("id", "")) if objective else "",
            "title": str((objective or {}).get("title", "")) if objective else "",
            "progress": _round((objective or {}).get("progress", 100), 1) if objective else 100,
            "hint": str(gameplay.get("objectiveHint", "")),
        },
        "gameplay": {
            "prosperity": _round(gameplay.get("prosperity", 0), 2),
            "threat": _round(gameplay.get("threat", 0), 2),
            "recovery": {
                "charges": int((gameplay.get("recovery") or {}).get("charges", 0) or 0),
                "activeBoostSec": _round(
                    (gameplay.get("recovery") or {}).get("activeBoostSec", 0), 1
                ),
                "collapseRisk": _round(
                    (gameplay.get("recovery") or {}).get("collapseRisk", 0), 1
                ),
                "lastReason": str((gameplay.get("recovery") or {}).get("lastReason", "")),
            },
        },
        "resources": {
            "food": _round((state.get("resources") or {}).get("food", 0), 2),
            "wood": _round((state.get("resources") or {}).get("wood", 0), 2),
        },
        "population": {"byGroup": population_by_group},
        "deaths": {
            "total": int(metrics.get("deathsTotal", 0) or 0),
            "byReason": dict(deaths),
        },
        "logistics": {
            "carryingWorkers": int(logistics.get("carryingWorkers", 0) or 0),
            "totalCarryInTransit": _round(logistics.get("totalCarryInTransit", 0), 2),
            "avgDepotDistance": _round(logistics.get("avgDepotDistance", 0), 2),
            "strandedCarryWorkers": int(logistics.get("strandedCarryWorkers", 0) or 0),
            "overloadedWarehouses": int(logistics.get("overloadedWarehouses", 0) or 0),
            "busiestWarehouseLoad": int(logistics.get("busiestWarehouseLoad", 0) or 0),
            "stretchedWorksites": int(logistics.get("stretchedWorksites", 0) or 0),
            "isolatedWorksites": int(logistics.get("isolatedWorksites", 0) or 0),
            "warehouseLoadByKey": dict(logistics.get("warehouseLoadByKey") or {}),
            "summary": str(logistics.get("summary", "Logistics: idle")),
        },
        "ecology": {
            "activeGrazers": int(ecology.get("activeGrazers", 0) or 0),
            "pressuredFarms": int(ecology.get("pressuredFarms", 0) or 0),
            "maxFarmPressure": _round(ecology.get("maxFarmPressure", 0), 2),
            "frontierPredators": int(ecology.get("frontierPredators", 0) or 0),
            "migrationHerds": int(ecology.get("migrationHerds", 0) or 0),
            "hotspotFarms": list(ecology.get("hotspotFarms") or []),
            "summary": str(ecology.get("summary", "Ecology: idle")),
            "activePressure": _round(
                max(
                    float(ecology.get("maxFarmPressure", 0) or 0),
                    float(spatial.get("eventPressure", 0) or 0),
                    float(spatial.get("weatherPressure", 0) or 0),
                ),
                2,
            ),
        },
        "warnings": {
            "count": len(metrics.get("warningLog") or []),
            "errorCount": len(error_warnings),
            "errorWarnings": error_warnings,
        },
        "performance": {
            "fps": _round(metrics.get("averageFps", 0), 2),
            "frameMs": _round(metrics.get("frameMs", 0), 2),
            "heapMb": _round(metrics.get("memoryMb", 0), 2),
            "renderMode": str((state.get("debug") or {}).get("renderMode", "unknown")),
            "entityCount": int(
                (metrics.get("populationStats") or {}).get(
                    "totalEntities",
                    len(state.get("agents") or []) + len(state.get("animals") or []),
                )
                or 0
            ),
            "renderFrameCount": int(metrics.get("renderFrameCount", 0) or 0),
            "simStepsThisFrame": int(metrics.get("simStepsThisFrame", 0) or 0),
            "topSystemMs": _build_top_system_ms(state, 3),
            "uiCpuMs": _round(metrics.get("uiCpuMs", 0), 2),
            "renderCpuMs": _round(metrics.get("renderCpuMs", 0), 2),
        },
        "ai": {
            "enabled": bool((state.get("ai") or {}).get("enabled")),
            "coverageTarget": str((state.get("ai") or {}).get("coverageTarget", "fallback")),
            "mode": str((state.get("ai") or {}).get("mode", "fallback")),
            "runtimeProfile": str((state.get("ai") or {}).get("runtimeProfile", "default")),
            "latencyMs": _round(metrics.get("aiLatencyMs", 0), 2),
            "fallbackActive": str((state.get("ai") or {}).get("mode", "fallback")) != "llm",
            "requestCount": int(ai_runtime.get("requestCount", 0) or 0),
            "timeoutCount": int(ai_runtime.get("timeoutCount", 0) or 0),
            "fallbackCount": int(ai_runtime.get("fallbackResponseCount", 0) or 0),
            "llmCount": int(ai_runtime.get("llmResponseCount", 0) or 0),
            "avgLatencyMs": _round(ai_runtime.get("avgLatencyMs", 0), 2),
            "consecutiveFallbackResponses": int(
                ai_runtime.get("consecutiveFallbackResponses", 0) or 0
            ),
            "maxUnrecoveredFallbackSec": _round(
                ai_runtime.get("maxUnrecoveredFallbackSec", 0), 2
            ),
            "recoveryCount": int(ai_runtime.get("recoveryCount", 0) or 0),
            "lastErrorKind": str(ai_runtime.get("lastErrorKind", "none")),
            "liveCoverageSatisfied": bool(ai_runtime.get("liveCoverageSatisfied", False)),
        },
        "view": {
            "targetX": _round((view_state or {}).get("targetX", 0), 2),
            "targetZ": _round((view_state or {}).get("targetZ", 0), 2),
            "zoom": _round((view_state or {}).get("zoom", 0), 3),
        },
        "actionMessage": str((state.get("controls") or {}).get("actionMessage", "")),
    }
    telemetry["nonFiniteMetrics"] = _collect_non_finite_metrics(telemetry)
    return telemetry
